pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -744,6 +792,117 @@ func applyLoadBalancer(cluster *apiv2.Cluster, lb *networking.LoadBalancerSettin
+		return
+	}
+
+	// one of Distribute or Failover settings can be applied.
+	if localityLB.GetDistribute() != nil {
+		applyLocalityWeight(proxy, loadAssignment, localityLB.GetDistribute())
+	} else if localityPriority {
+		applyLocalityFailover(proxy, loadAssignment, localityLB.GetFailover())
+	}
+}
+
+func applyLocalityWeight(
+	proxy *model.Proxy,
+	loadAssignment *apiv2.ClusterLoadAssignment,
+	distribute []*meshconfig.LocalityLoadBalancerSetting_Distribute) {
+	if distribute == nil {
+		return
+	}
+
+	for _, localityWeightSetting := range distribute {
+		if localityWeightSetting != nil &&
+			util.LocalityMatch(&proxy.Locality, localityWeightSetting.From) {
+			misMatched := map[int]struct{}{}
+			for i := range loadAssignment.Endpoints {
+				misMatched[i] = struct{}{}
+			}
+			for locality, weight := range localityWeightSetting.To {
+				// index -> original weight
+				destLocMap := map[int]uint32{}
+				totalWeight := uint32(0)
+				for i, ep := range loadAssignment.Endpoints {
+					if _, exist := misMatched[i]; exist {
+						if util.LocalityMatch(ep.Locality, locality) {
+							delete(misMatched, i)
+							destLocMap[i] = ep.LoadBalancingWeight.Value
+							totalWeight += destLocMap[i]
+						}
+					}
+				}
+				// in case wildcard dest matching multi groups of endpoints
+				// the load balancing weight for a locality is divided by the sum of the weights of all localities
+				for index, originalWeight := range destLocMap {
+					weight := float64(originalWeight*weight) / float64(totalWeight)
+					loadAssignment.Endpoints[index].LoadBalancingWeight = &types.UInt32Value{
+						Value: uint32(math.Ceil(weight)),
+					}
+				}
+			}
+
+			// remove groups of endpoints in a locality that miss matched
+			for i := range misMatched {
+				loadAssignment.Endpoints[i].LbEndpoints = nil
+			}
+			break
+		}
+	}
+}
+
+func applyLocalityFailover(
+	proxy *model.Proxy,
+	loadAssignment *apiv2.ClusterLoadAssignment,
+	failover []*meshconfig.LocalityLoadBalancerSetting_Failover) {
+	priorityMap := map[int][]int{}
+
+	for i, localityEndpoint := range loadAssignment.Endpoints {
+		priority := util.LbPriority(&proxy.Locality, localityEndpoint.Locality)
+		// region not match, apply failover settings
+		if priority == 3 {
+			for _, failoverSetting := range failover {
+				if failoverSetting.From == proxy.Locality.Region {
+					if localityEndpoint.Locality.Region != failoverSetting.To {
+						priority = 4
+					}
+					break
+				}
+			}
+		}
+		loadAssignment.Endpoints[i].Priority = uint32(priority)
+		priorityMap[priority] = append(priorityMap[priority], i)
+	}
+
+	// sort all priorities
+	priorities := []int{}
+	for priority := range priorityMap {
+		priorities = append(priorities, priority)
+	}
+	sort.Ints(priorities)
+	// adjust LocalityLbEndpoints priority
+	for i, priority := range priorities {
+		if i != priority {
+			for index := range priorityMap[priority] {
+				loadAssignment.Endpoints[index].Priority = uint32(i)
+			}
```

Comment:
why do you need this logic? isnt it sufficient to assign in the for loop (874) above?

Reply:
If above  set priority 1,3,4,  should adjust 3,4 to 2,3.

>// Priorities should range from 0 (highest) to N (lowest) without skipping.